pvoutput/grid_search.py

```diff
@@ -0,0 +1,763 @@
+    search_radius: Optional[float] = None,
+    local_crs_epsg: int = 4087,
+) -> Union[pd.DataFrame, gpd.GeoDataFrame]:
+    """Clip coordinates to a radius.
+
+    Remove any coordinates which do not lie within x.
+
+    Args:
+        coords:
+            A pandas DataFrame or geopandas GeoDataFrame of coordinates with columns: latitude,
+            longitude.
+        latitude:
+            Latitude of the center of the radial search.
+        longitude:
+            Longitude of the center of the radial search.
+        radius:
+            Set the radial search limit in km.
+        search_radius:
+            Optionally set the radial search limit around each grid point in kilometers. If set, the
+            code will consider coords to be included if any part of the search radius overlaps the
+            outter radius.
+        local_crs_epsg:
+            Optionally provide the EPSG code of a local Co-ordinate Reference System (CRS) for
+            improved accuracy. e.g. set to 27700 (OSGB36 / British National Grid) if searching the
+            British Isles.
+
+    Returns:
+        As per `coords` but containing only the subset of the input coordinates which fall within
+        `radius` km of the lat/lon.
+    """
+    center = Point(longitude, latitude)
+    radius_ = (
+        gpd.GeoSeries(center)
+        .set_crs("EPSG:4326")
+        .to_crs(f"EPSG:{local_crs_epsg}")
+        .buffer(radius * 1000.0)[0]
+    )
+    if search_radius is None:
+        coords["selected"] = coords_.within(radius_)
+    else:
+        coords_ = coords.to_crs(f"EPSG:{local_crs_epsg}").buffer(search_radius * 1000.0)
+        coords["selected"] = coords_.intersects(radius_)
+    return coords.loc[coords["selected"]].drop(columns="selected")
+
+
+def query_yes_no(question: str, default: Optional[str] = "yes") -> bool:
+    """Ask a yes/no question via input() and return the answer as boolean.
+
+    Args:
+        question:
+            The question presented to the user.
+        default:
+            The presumed answer if the user just hits <Enter>. It must be "yes" (the default), "no"
+            or None (meaning an answer is required of the user).
+
+    Returns:
+        Return value is True for "yes" or False for "no".
+    """
+    valid = {"yes": True, "y": True, "ye": True, "no": False, "n": False}
+    if default is None:
+        prompt = " [y/n] "
+    elif default == "yes":
+        prompt = " [Y/n] "
+    elif default == "no":
+        prompt = " [y/N] "
+    else:
+        raise ValueError("invalid default answer: '%s'" % default)
+    while True:
+        sys.stdout.write(question + prompt)
+        choice = input().lower()
+        if default is not None and choice == "":
+            return valid[default]
+        elif choice in valid:
+            return valid[choice]
+        else:
+            sys.stdout.write("Please respond with 'yes' or 'no' " "(or 'y' or 'n').\n")
+
+
+def parse_options():
```

Comment:
have you looked at click - its another package to take inputs in, and is quite nice. See example [here](https://github.com/openclimatefix/nowcasting_dataset/blob/main/scripts/prepare_ml_data.py

Reply:
I haven't although I just read up on it now. I must admit I wasn't completely sold! Is there a good argument for adding click as an additional requirement instead of using built-in argparse?

The docs mention:

> Why not Argparse?
> Click is internally based on optparse instead of argparse. This is an implementation detail that a user does not have to be concerned with. Click is not based on argparse because it has some behaviors that make handling arbitrary command line interfaces hard:
> 
> argparse has built-in behavior to guess if something is an argument or an option. This becomes a problem when dealing with incomplete command lines; the behaviour becomes unpredictable without full knowledge of a command line. This goes against Click’s ambitions of dispatching to subparsers.
> 
> argparse does not support disabling interspersed arguments. Without this feature, it’s not possible to safely implement Click’s nested parsing.

I don't think this is likely to affect this project though so perhaps the increased complexity / dependencies don't outweigh the benefits?

(Having said that, I can appreciate that consistency across the repo and across projects is desirable)

Reply:
No worries, leave it how it is . Thanks for looking into this thought